src/icmp.rs

```diff
@@ -0,0 +1,353 @@
+                0x0F => Self::DestinationUnreachable(Unreachable::PrecedentCutoffInEffect),
+                _ => Self::Other(raw),
+            },
+            0x04 => match c {
+                0x00 => Self::SourceQuench,
+                _ => Self::Other(raw),
+            },
+            0x05 => match c {
+                0x00 => Self::Redirect(Redirect::Network),
+                0x01 => Self::Redirect(Redirect::Host),
+                0x02 => Self::Redirect(Redirect::TosAndNetwork),
+                0x03 => Self::Redirect(Redirect::TosAndHost),
+                _ => Self::Other(raw),
+            },
+            0x07 => Self::Reserved,
+            0x08 => Self::EchoRequest,
+            0x09 => Self::RouterAdvertisment,
+            0x0A => Self::RouterSolicication,
+            0x0B => match c {
+                0x00 => Self::TimeExceeded(TimeExceeded::TTL),
+                0x01 => Self::TimeExceeded(TimeExceeded::FragmentReassembly),
+                _ => Self::Other(raw),
+            },
+            0x0C => match c {
+                0x00 => Self::ParameterProblem(ParameterProblem::Pointer),
+                0x01 => Self::ParameterProblem(ParameterProblem::MissingRequiredOption),
+                0x02 => Self::ParameterProblem(ParameterProblem::BadLength),
+                _ => Self::Other(raw),
+            },
+            0x0D => Self::Timestamp,
+            0x0E => Self::TimestampReply,
+            0x2A => Self::ExtendedEchoRequest,
+            0x2B => match c {
+                0x00 => Self::ExtendedEchoReply(ExtendedEchoReply::NoError),
+                0x01 => Self::ExtendedEchoReply(ExtendedEchoReply::MalformedQuery),
+                0x02 => Self::ExtendedEchoReply(ExtendedEchoReply::NoSuchInterface),
+                0x03 => Self::ExtendedEchoReply(ExtendedEchoReply::NoSuchTableEntry),
+                0x04 => Self::ExtendedEchoReply(ExtendedEchoReply::MupltipleInterfacesStatisfyQuery),
+                _ => Self::Other(raw),
+            }
+            _ => Self::Other(raw),
+        }
+    }
+}
+
+fn parse_icmp_code(input: &[u8]) -> IResult<&[u8], IcmpCode> {
+    let (input, code) = number::streaming::be_u16(input)?;
+
+    Ok((input, code.into()))
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum IcmpData {
+    Unreachable { nexthop_mtu: u16, header: IPv4Header, packet: [u8; 8] },
+    Redirect {gateway: Ipv4Addr, header: IPv4Header, packet: [u8; 8] },
+    TimeExceeded {header: IPv4Header, packet: [u8; 8] },
+    None,
+}
+
+fn parse_ipv4_header_and_packet(input: &[u8]) -> IResult<&[u8], (IPv4Header, [u8; 8])> {
+    let (input, header) = parse_ipv4_header(input)?;
+    if input.len() < 8 {
+        return Err(Err::Incomplete(Needed::Size(8 - input.len())));
+    }
+
+    let mut packet: [u8; 8] = Default::default();
+    packet.copy_from_slice(&input[..8]);
+    Ok((&input[8..], (header, packet)))
```

Comment:
https://docs.rs/nom/6.2.1/nom/bytes/streaming/fn.take.html

Reply:
Thanks,
Fixed in https://github.com/bestouff/pktparse-rs/pull/12/commits/63056f68094a20de49f9d9b44600cf36b0401e86